clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+
+    Params.push_back(NewParam);
+
+    // Create the new type.
+    ArgTys.push_back(ArgTy);
+    FunctionProtoType::ExtProtoInfo Info(CC_OpenCLKernel);
+    QualType FuncType = Ctx.getFunctionType(Ctx.VoidTy, ArgTys, Info);
+
+    // Set type, note until the destructor of the owner of KernelObj is called
+    // (to set the parameters), we cannot access the parameters from KernelObj
+    // without memory problems.
+    KernelObj->setType(FuncType);
+  }
+
+  // All special SYCL objects must have __init method. We extract types for
+  // kernel parameters from __init method parameters. We will use __init method
+  // and kernel parameters which we build here to initialize special objects in
+  // the kernel body.
+  void handleSpecialType(const FieldDecl *FD, QualType ArgTy) {
+    const auto *RecordDecl = ArgTy->getAsCXXRecordDecl();
+    assert(RecordDecl && "Special SYCL object must be of a record type");
+
+    CXXMethodDecl *InitMethod = getMethodByName(RecordDecl, InitMethodName);
+    assert(InitMethod && "The accessor/sampler must have the __init method");
+
+    for (const ParmVarDecl *Param : InitMethod->parameters())
+      addParam(FD, Param->getType().getCanonicalType());
+  }
+
+public:
+  SyclKernelDeclBase(Sema &S, FunctionDecl *KernelObj,
+                     llvm::SmallVectorImpl<QualType> &ArgTys,
+                     llvm::SmallVectorImpl<ParmVarDecl *> &Params)
+      : SyclKernelFieldHandler(S), KernelObj(KernelObj), ArgTys(ArgTys),
+        Params(Params) {}
+
+  void handleSyclAccessorType(const FieldDecl *FD, QualType ArgTy) final {
+    handleSpecialType(FD, ArgTy);
+  }
+
+  void handleStructType(const FieldDecl *FD, QualType ArgTy) override {
+    addParam(FD, ArgTy);
+
+    // Create descriptors for each accessor field in the class or struct
+    const auto *Wrapper = ArgTy->getAsCXXRecordDecl();
+    VisitKernelFields(Wrapper->fields(), *this);
+  }
+};
+
+// A type to Create and own the FunctionDecl for the kernel.
+class SyclKernelDeclCreator : public SyclKernelDeclBase {
+  // TODO: rather than this, should we consider a 'commit' function that
+  // finalizes under success only?
+  SyclKernelFieldChecker &ArgChecker;
+  Sema::ContextRAII FuncContext;
+  // Yes, the list of Parameters contains this info, but we use it often enough
+  // we shouldn't be recreating it constantly.  QualTypes are cheap anyway.
+  llvm::SmallVector<QualType, 8> ArgTys;
+  llvm::SmallVector<ParmVarDecl *, 8> Params;
+
+  static void setKernelImplicitAttrs(ASTContext &Context, FunctionDecl *FD,
+                                     StringRef Name) {
+    // Set implict attributes.
+    FD->addAttr(OpenCLKernelAttr::CreateImplicit(Context));
+    FD->addAttr(AsmLabelAttr::CreateImplicit(Context, Name));
+    FD->addAttr(ArtificialAttr::CreateImplicit(Context));
+  }
+
+  static FunctionDecl *initKernelObj(ASTContext &Ctx, StringRef Name,
+                                     SourceLocation Loc, bool IsInline) {
```

Comment:
I don't get why it is called `initKernelObj`. Shouldn't it be something like `createKernelDecl`?
BTW, I don't think that OpenCL kernel somewhat can be inline. o_O

Reply:
The implicit inline logic came from the existing implementation.

Naming things is hard... 'initKernelObj' (with the variable named KernelObj) made sense at the time *shrug*.

Reply:
I think because the name `KernelObj` is incorrect.
```
FunctionDecl *KernelObj;
```
I think this variable contains declaration (and definition after we generate the body) of OpenCL kernel. We have to rename it so it will be easy to understand that this is OpenCL kernel function declaration. Something like `KernelDecl`.